Added construction preprocess endpoint.

cngo-github opened the pull request:
__PR Category__: Feature

Added Rosetta's [`construction/preprocess` endpoint](https://www.rosetta-api.org/docs/ConstructionApi.html#constructionpreprocess).

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Amount.scala:
```diff
@@ -0,0 +1,55 @@
+    implicit val amountValueRefinedEncoder: Encoder[AmountValueRefined] =
+      encoderOf[Long, Not[0L]]
+
+    implicit val amountValueRefinedDecoder: Decoder[AmountValue] =
```
I am working on getting the `Decoder` up and running. I tried `Validate.fromPredicate`, but I am not sure what the third argument should be. Pointers here are greatly appreciated.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/api/construction/ConstructionPreprocess.scala:
```diff
@@ -0,0 +1,24 @@
+  case class Request(
+    networkIdentifier: NetworkIdentifier,
+    operations: List[Operation],
+    maxFee: Option[List[Amount]],
+    suggestedFeeMultiplier: Option[Double]
```
can you show me these properties in the rosetta specification? I cannot find any on the preprocess request: 
https://www.rosetta-api.org/docs/models/ConstructionPreprocessRequest.html

cngo-github replied:
I just copied the class over and didn't think to consult the specification. That's on me. Removed.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -39,5 +41,12 @@ object ConstructionService {
       .attemptT
       .leftMap(_ => MalformedTransaction)
+
+    def getAccountIdentifiers(operations: List[Operation]): Option[List[AccountIdentifier]] = {
+      val negativeValueOperations = operations.filter(_.amount.exists(_.value < 0))
+
+      val test = Some(negativeValueOperations.flatMap(_.account)).filter(test => test.nonEmpty)
+      test
+    }
```
we've already rewritten that on slack, please take a look there

cngo-github replied:
That code depended on `amount` being an `Option[Either[NegAmount, PosAmount]]`. It is now just a `Long Refined Not[Equal[0L]]`. I'll play around with this to see if I can make something cleaner.

cngo-github replied:
This is what I came up with

```
def getAccountIdentifiers(operations: List[Operation]): Option[NonEmptyList[AccountIdentifier]] = {
      val accountIdentifiers =
        operations.filter(_.amount.exists(_.value.value.value < 0)).flatMap(_.account)

      Option.when(accountIdentifiers.nonEmpty)(NonEmptyList.fromList(accountIdentifiers)).flatten
    }
```

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Amount.scala:
```diff
@@ -0,0 +1,55 @@
+/** Rosetta Build Once. Integrate Your Blockchain Everywhere.
+  *
+  * The version of the OpenAPI document: 1.4.12 Contact: team@openapitools.org
+  *
+  * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech). https://openapi-generator.tech
+  */
+
```
Oversight. It has been removed.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Currency.scala:
```diff
@@ -0,0 +1,25 @@
+  case class Currency(
+    symbol: CurrencySymbol,
+    decimals: NonNegLong
+  )
```
So, I can have the types and and the other case classes in an object called `currency` and not have to think about another name.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Operation.scala:
```diff
@@ -0,0 +1,49 @@
+  case class Operation(
+    operationIdentifier: OperationIdentifier,
+    relatedOperations: Option[List[OperationIdentifier]],
+    `type`: TransferOperation,
```
Changed.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/http/routes/ConstructionRoutes.scala:
```diff
@@ -36,6 +34,14 @@ final class ConstructionRoutes[F[_]: Async](
+      req.decodeRosettaWithNetworkValidation[ConstructionPreprocess.Request](appEnvironment, _.networkIdentifier) { preprocessReq =>
+        constructionService
+          .getAccountIdentifiers(preprocessReq.operations)
+          .map(accountIds => ConstructionPreprocess.Response(Some(accountIds)))
```
The type here is `Option[ConstructionPreprocess.Response]`. How do I convert that to `F[Response[F]]` to then apply `.asRosettaResponse.handleUnknownError`?

marcinwadon replied:
so, in this case there is no error returned by the service (as the function is pure) so you can forget about the error and just return 200 Ok

```scala
{ preprocessReq =>
  val accountIdentifiers = constructionService.getAccountIdentifiers(preprocessReq.operations)
  
  Ok(ConstructionPreprocess.Response(accountIdentifiers))
}
```
or
```scala
{ preprocessReq =>
  val accountIdentifiers = constructionService.getAccountIdentifiers(preprocessReq.operations)
  
  Ok(ConstructionPreprocess.Response(accountIdentifiers))
    .handleUnknownError
}
```

marcinwadon replied:
I cannot confirm it right now, but if above would not be sufficient (by returning `Response[F]` instead of `F[Reponse[F]]`), you can always wrap it into `F` by `.pure[F]` (from `Applicative`)

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Currency.scala:
```diff
@@ -0,0 +1,29 @@
+  @newtype
+  case class CurrencySymbol(value: CurrencySymbolRefined)
+
+  @derive(decoder, encoder)
+  @newtype
+  case class CurrencyDecimal(value: NonNegLong)
```
should we bind it to 8 decimals for our network?

cngo-github replied:
That decision is beyond my pay grade. I don't know the network well enough. I'll leave that to you guys and @ryle-goehausen to decide.

cngo-github replied:
Refined the value to 8.

```
type CurrencyDecimalRefined = Long Refined Equal[8L]

  @derive(decoder, encoder)
  @newtype
  case class CurrencyDecimal(value: CurrencyDecimalRefined)
```

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Operation.scala:
```diff
@@ -0,0 +1,51 @@
+  object OperationStatus extends StringEnum[OperationStatus] with StringCirceEnum[OperationStatus] {
+    val values = findValues
+
+    case object Accepted extends OperationStatus(value = "accepted")
+    case object Pending extends OperationStatus(value = "pending")
+    case object Rejected extends OperationStatus(value = "rejected")
+    case object Unknown extends OperationStatus(value = "unknown")
```
![image](https://user-images.githubusercontent.com/4959931/218938592-e91cd133-6609-460c-a664-6a44067269c5.png)

Rosetta uses camel case for both `OperationStatus` and `OperationType`.
if this is network dependant and we can use any convention we want, I'd keep it unified.
see `accepted` and `TRANSFER`

cngo-github replied:
Standardized on camelcase.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -39,5 +41,12 @@ object ConstructionService {
+      val accountIdentifiers =
+        operations.filter(_.amount.exists(_.value.value.value < 0)).flatMap(_.account)
+
+      Option.when(accountIdentifiers.nonEmpty)(NonEmptyList.fromList(accountIdentifiers)).flatten
```
if you decided to use `NonEmptyList.fromList` it already returns an Option.
None in case the list is empty and Some in case there are elements, so there is no need to wrap it additionally into second Option and flatten instantly.

so it should be just:
```scala
NonEmptyList.fromList(accountIdentifiers)
```

cngo-github replied:
Changed.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -39,5 +41,12 @@ object ConstructionService {
+
+    def getAccountIdentifiers(operations: List[Operation]): Option[NonEmptyList[AccountIdentifier]] = {
+      val accountIdentifiers =
+        operations.filter(_.amount.exists(_.value.value.value < 0)).flatMap(_.account)
```
`_.value.value.value` looks weird and I'd think about supportive functions in `AmountValue` like `def isNegative`/`def isPositive`. Then, this would look like `_.amount.exists(_.value.isNegative)`

cngo-github replied:
Added `isNegative` to `AmountValue`.

marcinwadon reviewed (changes requested):
Overall looks great, good job :)

cngo-github commented:
I will add the tests, then this PR will be ready to be removed from `draft` status.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/Amount.scala:
```diff
@@ -0,0 +1,28 @@
+  @derive(decoder, encoder)
+  @newtype
+  case class AmountValue(value: AmountValueRefined) {
+    def isNegative: Boolean = value.value < 0
```
try this, it should work with no unwrapping:
```scala
def isNegative: Boolean = value < 0L
```

kpudlik replied:
eventually try adding `import io.estatico.newtype.ops._` if it won't work

cngo-github replied:
By changing `value.value` to `value`, I get `value < is not a member of org.tessellation.rosetta.domain.amount.AmountValueRefine`. I am importing `io.estatico.newtype.ops._`.

marcinwadon replied:
try with `import eu.timepit.refined.auto._` or `import eu.timepit.refined.cats._`.
it works on my side

marcinwadon replied:
yes, this is the correct one `import eu.timepit.refined.auto._`

cngo-github replied:
Changed. I need help with fixing my test now.

marcinwadon reviewed (dismissed):
:clap:

cngo-github commented on modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:
```diff
@@ -109,4 +114,64 @@ object ConstructionServiceSuite extends MutableIOSuite with Checkers with Transa
+              Transfer,
+              None,
+              Some(AccountIdentifier(address, None)),
+              Some(Amount(AmountValue(1L), Currency(CurrencySymbol("DAG"), CurrencyDecimal(1L))))
```
Nevermind on this. I didn't really understand `for` comprehension.

marcinwadon commented on modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:
```diff
@@ -109,4 +114,66 @@ object ConstructionServiceSuite extends MutableIOSuite with Checkers with Transa
+
+    val cs = ConstructionService.make[IO]()
+
+    val gen = for {
+      address <- addressGen
+    } yield address
```
for comprehension is a syntax sugar for flatMap.
```
val gen = for {
  address <- addressGen
  } yield address
```
is equal to
```
val gen = addressGen
```

so there is no point to write a for comp. for one operation.

anyway, if you have more complex case, you can use both `<-` and `=` as follows:
```scala
for {
 lorem1 <- ipsum1()
 lorem2 <- ipsum2()
 lorem3 = ipsum3()
 } yield (lorem1, lorem2, lorem3)
```
is equal to:
```
ipsum1().flatMap { lorem1 =>
  ipsum2().map { lorem2 =>
    val lorem3 = ipsum3()
    
    (lorem1, lorem2, lorem3)
  }
}
```

cngo-github replied:
Removed the unnecessary `for` comprehension

marcinwadon commented on modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:
```diff
@@ -109,4 +114,66 @@ object ConstructionServiceSuite extends MutableIOSuite with Checkers with Transa
+      address <- addressGen
+    } yield address
+
+    forall(gen) {
+      case address =>
+        val result = for {
+          operation <- Some(
+            Operation(
+              OperationIdentifier(OperationIndex(1L)),
+              None,
+              Transfer,
+              None,
+              Some(AccountIdentifier(address, None)),
+              Some(Amount(AmountValue(-1L), Currency(CurrencySymbol("DAG"), CurrencyDecimal(8L))))
+            )
+          )
+          accountIdentifiers <- cs.getAccountIdentifiers(operation :: Nil)
+          extractedAddress <- Some(accountIdentifiers.map(id => id.address).head)
+        } yield expect.same(extractedAddress, address)
+
+        result.getOrElse(failure("The addresses are not the same."))
+    }
```
1. there is no need to wrap everything into `Some` in the for comprehension. getAccountIdentifiers is a pure function, so it can be simply assigned to `val` or `def`
2. please don't use `expect.same` even if you see such examples in the codebase. we should use `expect.eql` which does not import implicit universal equal (and we are going to change all `expect.same` usages into `expect.eql` in the existing code)
3. in 99% cases you don't need to return `failure` on your own. you wrote an expectation that addresses should be the same, so use this expectation. it will return success or failure automatically

```scala
forall(addressGen) {
  case address =>
    val operation = Operation(
      OperationIdentifier(OperationIndex(1L)),
      None,
      Transfer,
      None,
      AccountIdentifier(address, None).some,
      Amount(AmountValue(-1L), Currency(CurrencySymbol("DAG"), CurrencyDecimal(8L))).some
    )
    
    val result = cs.getAccountIdentifiers(List(operation)).map(_.address).head
    
    expect.eql(address, result)
}
```

cngo-github replied:
The `.getOrElse(failure("The addresses are not the same."))` is there because the type of `result` is `Option[Expectation]`. To unwrap an `Option`, I only know of `.get` and `.getOrElse` and `.getOrElse` is "less bad". The rest has been cleaned up.

marcinwadon replied:
I talked with @kpudlik and basically instead of unwrapping result, you can lift an expected address to Option

kpudlik replied:
```
    forall(addressGen) {
      case address =>
        val operation = Operation(
          OperationIdentifier(OperationIndex(1L)),
          None,
          Transfer,
          None,
          Some(AccountIdentifier(address, None)),
          Some(Amount(AmountValue(-1L), Currency(CurrencySymbol("DAG"), CurrencyDecimal(8L))))
        )

        val expectedAddress: Option[Address] = cs.getAccountIdentifiers(operation :: Nil)
          .map(_.head.address)

        expect.eql(expectedAddress, Some(address)) // You can't compare `Address` with `Option[Address]` but you can lift `Address` to `Option` (wrap by effect so wrap by `Option`) and then you will compare `Option[Address]` with `Option[Address]`
    }
```

marcinwadon replied:
@cngo-github I see that you didn't change the rest of the things, especially you still use `expect.same`.

```scala
  test("returns the accountIdentifiers for negative operations") { res =>
    implicit val (sp, k) = res

    val cs = ConstructionService.make[IO]()

    forall(addressGen) { address =>
      val operation = Operation(
        OperationIdentifier(OperationIndex(1L)),
        none,
        Transfer,
        none,
        AccountIdentifier(address, none).some,
        Amount(AmountValue(-1L), Currency(CurrencySymbol("DAG"), CurrencyDecimal(8L))).some
      )

      val result = cs
        .getAccountIdentifiers(List(operation))
        .map(_.head.address)

      expect.eql(address.some, result)
    }
  }

```

cngo-github replied:
I will need guidance on this.

This

```
cs.getAccountIdentifiers(operation :: Nil)
          .map(_.head.address).liftTo[Option]
```

gives me

```
OptionOps.LiftToPartiallyApplied[Option, Address]
```

and I don’t know where to go from there to make it an `Expectation`.

marcinwadon replied:
please take a look at @kpudlik comment. there is an annotation in the snippet

cngo-github replied:
Changed.

marcinwadon reviewed (changes requested):
As previous

cngo-github commented:
> As previous

- Removed the `case` for both tests.
- Re-arranged the ordering for `expect.eql` on the negative operations test.
- Changed `Some(address)` to `address.some`.

cngo-github commented:
> As previous

- `operation :: Nil` changed to `List(Operation)`.

cngo-github commented:
- Changed all `None` to `none`.
- Changed all `Some(…)` to `.some`.
